test/test_type_promotion.py

```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
         # this seems like odd behavior but ints also create float tensors, numpy doesn't have this function.
         self.assertEqual(torch.scalar_tensor(False, device=device), torch.tensor(0., device=device))
 
+    @dtypes(*itertools.product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes()))
+    def test_result_type(self, device, dtypes):
+        "Test result_type for tensor vs tensor and scalar vs scalar."
+
+        def _get_dtype(x):
+            "Get the dtype of x if x is a tensor. If x is a scalar, get its corresponding dtype if it were a tensor."
+            if torch.is_tensor(x):
+                return x.dtype
+            elif isinstance(x, bool):
+                return torch.bool
+            elif isinstance(x, int):
+                return torch.int64
+            elif isinstance(x, float):
+                return torch.float32
+            elif isinstance(x, complex):
+                return torch.complex64
+            else:
+                raise AssertionError(f"Unkonwn type {x}")
+
+        # tensor against tensor
+        a_tensor = torch.tensor((0, 1), device=device, dtype=dtypes[0])
+        a_single_tensor = torch.tensor(1, device=device, dtype=dtypes[0])
+        a_scalar = a_single_tensor.item()
+        b_tensor = torch.tensor((1, 0), device=device, dtype=dtypes[1])
+        b_single_tensor = torch.tensor(1, device=device, dtype=dtypes[1])
+        b_scalar = b_single_tensor.item()
+        for a, b in ((a_tensor, b_tensor), (a_single_tensor, b_single_tensor), (a_scalar, b_scalar)):
+            dtype_a = _get_dtype(a)
+            dtype_b = _get_dtype(b)
+            try:
+                result = a + b
+            except RuntimeError:
+                with self.assertRaises(RuntimeError):
+                    torch.promote_types(dtype_a, dtype_b)
+                with self.assertRaises(RuntimeError):
+                    torch.result_type(a, b)
+            else:
+                dtype_res = _get_dtype(result)
+                if a is a_scalar and b is b_scalar and dtype_a == torch.bool and dtype_b == torch.bool:
+                    # special case: in Python, True + True is an integer
+                    self.assertEqual(dtype_res, torch.int64, f"a == {a}, b == {b}")
+                else:
+                    self.assertEqual(dtype_res, torch.result_type(a, b), f"a == {a}, b == {b}")
+                if a is a_scalar and b is b_scalar:  # Python internal type determination is good enough in this case
+                    continue
+                self.assertEqual(dtype_res, torch.promote_types(dtype_a, dtype_b), f"a == {a}, b == {b}")
```

Comment:
I don't think this assertion will be true if this test includes tensor x scalar tests.

Reply:
One option is to only test `torch.promote_types` when both arguments are 1D+ tensors. Then this wouldn't need `_get_dtype`, either. (The test could also validate that `torch.promote_types` only throws an error if both arguments were 1D+ tensors, too.)

Reply:
I have enclosed this line in a very simple `if` block. It would not decrease the coverage of this line at all, but also doesn't add complications.